### PR TITLE
Milestone 6: Testing ruby code - Enumerables

### PR DESCRIPTION
**Pull request by rajkumardas2701**
Created test cases for Enumerable methods

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Job guys :clap:
You have done a tremendous progress in this project. Most test cases have been covered however, we need to add a few cases especially for negative edge cases before we can merge this PR.
- check my comments for suggestions
- make your README more professional. How does someone pull the repo on their PC? add instructions for cloning and running test cases. Your README only has instruction for executing the main.rb file but remember we are testing here so add relevant instructions.

### Martin Afani

File: spec/main_spec.rb [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+  describe 'my_each' do
+    it 'should display all the numbers in an array' do
+      a = %w[sharon leo anthony]
+      expect(a.my_each { |x| x }).to eq(%w[sharon leo anthony])
+    end
+
+    it 'should return an enum if no block is given' do
+      a = %w[sharon leo anthony]
+      expect(a.my_each.class).to eq(Enumerator)
+    end
```
Comment (whiz25) [inline spec/main_spec.rb:13]: - [ ] what should be returned when an Enumerator is provided? Always include tests for +ve and -ve cases

File: spec/main_spec.rb [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    it 'should return an enum if no block is given' do
+      a = %w[sharon leo anthony]
+      expect(a.my_each.to_enum).to be_a(Enumerator)
+    end
+
+    it 'should return Class when no Enumerator and block given' do
+      a = %w[sharon leo anthony]
+      expect(a.my_each.to_enum).not_to eq(Class)
+    end
+  end
```
Comment (shshamim63) [inline spec/main_spec.rb:24]: Actually there is a better way to organize the test. Take a look at the following formation of code.
```ruby
describe '#my_each' do
    let(:arr) { [1, 2, 3] }
    context 'when no block is given' do
      it 'returns an enumerator' do
        expect(arr.my_each).to an_instance_of(Enumerator)
      end
      it 'returns same type value as the original each' do
        expect(arr.my_each.class).to eql(arr.each.class)
      end
    end
    context 'when a block is given' do
      it 'calls the block for each of the item in the enum and returns self' do
        expect(arr.my_each {|num| num > 5 }).to eq(arr)
      end
      it 'calls the block for each of the element in the enum as the original each' do
        each_output = 0
        arr.my_each {|num| each_output += num }
        my_each_output = each_output.dup
        each_output = 0
        arr.each {|num| each_output += num }
        expect(my_each_output).to eq(each_output)
      end
    end
end

```

**Review by shshamim63 — APPROVED** [review]
## Status :+1: Approved
___
Hi @rajkumardas2701 :wave: 
Nice job :clap: , You have added lot of tests, and the tests are functional. But there are some issues that the previous reviewer forgot to inform you(I am sorry for that :disappointed: ). So I am declaring them as optional changes.
___
### :clubs: Optional Changes :clubs: 
___
1. You can make your test more organized in several blocks. Check out the sample test I have given.
2. If you are using a set of values multiple times then it is better to set the value in a variable using let.
3. To make your tests super solid you can compare the expect() with the actual method.
___
##### This :+1:  is from me. You can go ahead. If you need more explanation don't hesitate to knock me on slack.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)
